scylla-cdc-replicator/src/replicator_consumer.rs

```diff
@@ -0,0 +1,261 @@
+        table_schema
+            .partition_key
+            .iter()
+            .chain(table_schema.clustering_key.iter())
+    }
+
+    fn bind_marker_for_type(val_type: &CqlType) -> String {
+        if let CqlType::Tuple(vec) = val_type {
+            format!(
+                "({})",
+                vec.iter()
+                    .map(ReplicatorConsumer::bind_marker_for_type)
+                    .join(",")
+            )
+        } else {
+            "?".to_string()
+        }
+    }
+
+    async fn update(&self, data: CDCRow<'_>) {
+        self.update_or_insert(data, false).await;
+    }
+
+    async fn insert(&self, data: CDCRow<'_>) {
+        self.update_or_insert(data, true).await;
+    }
+
+    async fn update_or_insert(&self, data: CDCRow<'_>, is_insert: bool) {
+        let keys_iter = ReplicatorConsumer::get_keys_iter(&self.table_schema).await;
+        let ttl = CqlValue::BigInt(data.ttl.unwrap_or(0)); // If data is inserted without TTL, setting it to 0 deletes existing TTL.
+        let timestamp = 0; // [TODO]: Map uuid to i64
+        let values = keys_iter
+            .clone()
+            .map(|col_name| data.get_value(col_name).as_ref().unwrap())
+            .collect::<Vec<&CqlValue>>();
+
+        if is_insert {
+            // Insert row with nulls, the rest will be done through an update.
+            self.run_prepared_statement(self.insert_query.clone(), &values, timestamp)
+                .await;
+        }
+
+        let mut values_for_update = vec![&ttl, &CqlValue::Int(0)];
+        values_for_update.reserve(values.len());
+        values_for_update.extend(values.iter());
+
+        for column_name in &self.other_columns {
+            if let Some(value) = data.get_value(column_name) {
+                // Order of values: ttl, inserted value, pk condition values.
+                values_for_update[1] = value;
+                // [TODO]: Precompute these queries.
+                self.run_statement(
+                    Query::new(format!(
+                        "UPDATE {}.{} USING TTL ? SET {} = {} WHERE {}",
+                        self.dest_keyspace_name,
+                        self.dest_table_name,
+                        column_name,
+                        ReplicatorConsumer::bind_marker_for_type(self.get_column_type(column_name)),
+                        self.keys_cond
+                    )),
+                    &values_for_update,
+                    timestamp,
+                )
+                .await;
+            } else if data.is_value_deleted(column_name) {
+                // [TODO]: Precompute these queries.
+                self.run_statement(
+                    Query::new(format!(
+                        "DELETE {} FROM {}.{} WHERE {}",
+                        column_name, self.dest_keyspace_name, self.dest_table_name, self.keys_cond
+                    )),
+                    &values,
+                    timestamp,
+                )
+                .await;
+            }
```

Comment:
This will work, but only for frozen collections/UDTs and atomic types. Please move this logic to a separate function. In this loop, please check the column's type using information in the `table_schema` and call the new function if it matches. This way it will be easier for others to add support for other types in parallel.